File: backend/src/modules/configuracion/configuracion-tienda.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '@common/config/database/prisma.service';
import { ConfigurarTiendaDto } from './dto/configurar-tienda.dto';

export interface ConfiguracionTiendaResponse {
  id: string;
  nombreTienda: string | null;
  whatsappContacto: string | null;
  banco: string | null;
  cbu: string | null;
  alias: string | null;
  titular: string | null;
  mensajeTransferencia: string | null;
  pedidoVencimientoHoras: number;
  estadoProductoBorrador: boolean;
  actualizadoEn: Date;
}

@Injectable()
export class ConfiguracionTiendaService {
  constructor(private readonly prisma: PrismaService) {}

  async obtener(): Promise<ConfiguracionTiendaResponse> {
    let configuracion = await this.prisma.configuracionTienda.findUnique({
      where: { id: 'global' },
    });

    if (!configuracion) {
      configuracion = await this.prisma.configuracionTienda.create({
        data: {
          id: 'global',
          nombreTienda: null,
          whatsappContacto: null,
          banco: null,
          cbu: null,
          alias: null,
          titular: null,
          mensajeTransferencia: null,
          pedidoVencimientoHoras: 48,
          estadoProductoBorrador: true,
        },
      });
    }

    return this.mapearRespuesta(configuracion);
  }

  async actualizar(dto: ConfigurarTiendaDto): Promise<ConfiguracionTiendaResponse> {
    const datosActualizacion: Record<string, unknown> = {};

    if (dto.nombreTienda !== undefined) datosActualizacion.nombreTienda = dto.nombreTienda;
    if (dto.whatsappContacto !== undefined) datosActualizacion.whatsappContacto = dto.whatsappContacto;
    if (dto.banco !== undefined) datosActualizacion.banco = dto.banco;
    if (dto.cbu !== undefined) datosActualizacion.cbu = dto.cbu;
    if (dto.alias !== undefined) datosActualizacion.alias = dto.alias;
    if (dto.titular !== undefined) datosActualizacion.titular = dto.titular;
    if (dto.mensajeTransferencia !== undefined) datosActualizacion.mensajeTransferencia = dto.mensajeTransferencia;
    if (dto.pedidoVencimientoHoras !== undefined) datosActualizacion.pedidoVencimientoHoras = dto.pedidoVencimientoHoras;
    if (dto.estadoProductoBorrador !== undefined) datosActualizacion.estadoProductoBorrador = dto.estadoProductoBorrador;

    const configuracion = await this.prisma.configuracionTienda.update({
      where: { id: 'global' },
      data: datosActualizacion,
    });

    return this.mapearRespuesta(configuracion);
  }

  private mapearRespuesta(
    config: {
      id: string;
      nombreTienda: string | null;
      whatsappContacto: string | null;
      banco: string | null;
      cbu: string | null;
      alias: string | null;
      titular: string | null;
      mensajeTransferencia: string | null;
      pedidoVencimientoHoras: number;
      estadoProductoBorrador: boolean;
      actualizadoEn: Date;
    },
  ): ConfiguracionTiendaResponse {
    return {
      id: config.id,
      nombreTienda: config.nombreTienda,
      whatsappContacto: config.whatsappContacto,
      banco: config.banco,
      cbu: config.cbu,
      alias: config.alias,
      titular: config.titular,
      mensajeTransferencia: config.mensajeTransferencia,
      pedidoVencimientoHoras: config.pedidoVencimientoHoras,
      estadoProductoBorrador: config.estadoProductoBorrador,
      actualizadoEn: config.actualizadoEn,
    };
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/configuracion/ConfiguracionTiendaService.java
package com.sistemapedidos.modules.configuracion;

import com.sistemapedidos.domain.entity.ConfiguracionTienda;
import com.sistemapedidos.domain.repository.ConfiguracionTiendaRepository;
import com.sistemapedidos.modules.configuracion.dto.ConfiguracionTiendaResponse;
import com.sistemapedidos.modules.configuracion.dto.ConfigurarTiendaDto;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class ConfiguracionTiendaService {

    public static final String ID_GLOBAL = "global";

    private final ConfiguracionTiendaRepository configuracionTiendaRepository;

    @Transactional
    public ConfiguracionTiendaResponse obtener() {
        ConfiguracionTienda configuracion = configuracionTiendaRepository.findById(ID_GLOBAL)
                .orElseGet(() -> {
                    ConfiguracionTienda nueva = new ConfiguracionTienda();
                    nueva.setId(ID_GLOBAL);
                    return configuracionTiendaRepository.save(nueva);
                });
        return mapearRespuesta(configuracion);
    }

    @Transactional
    public ConfiguracionTiendaResponse actualizar(ConfigurarTiendaDto dto) {
        ConfiguracionTienda configuracion = configuracionTiendaRepository.findById(ID_GLOBAL)
                .orElseGet(() -> {
                    ConfiguracionTienda nueva = new ConfiguracionTienda();
                    nueva.setId(ID_GLOBAL);
                    return nueva;
                });

        if (dto.nombreTienda() != null) configuracion.setNombreTienda(dto.nombreTienda());
        if (dto.whatsappContacto() != null) configuracion.setWhatsappContacto(dto.whatsappContacto());
        if (dto.banco() != null) configuracion.setBanco(dto.banco());
        if (dto.cbu() != null) configuracion.setCbu(dto.cbu());
        if (dto.alias() != null) configuracion.setAlias(dto.alias());
        if (dto.titular() != null) configuracion.setTitular(dto.titular());
        if (dto.mensajeTransferencia() != null) configuracion.setMensajeTransferencia(dto.mensajeTransferencia());
        if (dto.pedidoVencimientoHoras() != null) configuracion.setPedidoVencimientoHoras(dto.pedidoVencimientoHoras());
        if (dto.estadoProductoBorrador() != null) configuracion.setEstadoProductoBorrador(dto.estadoProductoBorrador());

        configuracionTiendaRepository.save(configuracion);
        return mapearRespuesta(configuracion);
    }

    private ConfiguracionTiendaResponse mapearRespuesta(ConfiguracionTienda config) {
        return new ConfiguracionTiendaResponse(
                config.getId(),
                config.getNombreTienda(),
                config.getWhatsappContacto(),
                config.getBanco(),
                config.getCbu(),
                config.getAlias(),
                config.getTitular(),
                config.getMensajeTransferencia(),
                config.getPedidoVencimientoHoras(),
                config.getEstadoProductoBorrador(),
                config.getActualizadoEn());
    }
}
